### ISSUE TITLE
Delete or Create key fails: Attempting to mutate more than 1 global secondary index at the same time

### ISSUE THREAD

**Issue by malcomm** [aws-amplify/amplify-cli#5380]
**Describe the bug**

When attempting to delete a single key (no other changes) I get:

```
Attempting to mutate more than 1 global secondary index at the same time on the FooTable table in the Foo stack. 
An error occurred during the push operation: Attempting to mutate more than 1 global secondary index at the same time on the FooTable table in the Foo stack. 
```

**Amplify CLI Version**

4.29.2

**To Reproduce**

1. Have a schema that has already been pushed with an example `@key`
2. Try to delete the `@key`
3. `amplify push` fails

**Expected behavior**

`@key` should be deleted

**Additional context**

Example schema:
```graphql
type Foo
  @model
  @key(name: "FooStatusUpdatedAt", fields: ["status", "updatedAt"], queryField: "fooStatusUpdatedAt")
  @searchable
{
  updatedAt: AWSDateTime,
  status: PublicationStatus,
  ...
}

enum PublicationStatus {
  draft
}
```

**Note:** I downgraded to `@aws-amplify/cli@4.27.3` and was able to perform the `amplify push` without issue.

**Comment by ammarkarachi** [comment]
@malcomm was the initial push with an older version?

**Comment by malcomm** [comment]
@ammarkarachi - no I just added this key today and pushed it.

**Comment by ammarkarachi** [comment]
I ended up using this query to add and delete the GSI with some data, I was able to  push to successfully in both instances. It could be the GSI was getting updated when the delete that being counted as another mutation. 
```graphql
type Foo
  @model
  @key(name: "FooStatusUpdatedAt", fields: ["status", "updatedAt"], queryField: "fooStatusUpdatedAt")
{
  updatedAt: AWSDateTime,
  status: PublicationStatus,
}

enum PublicationStatus {
  draft
}
```

**Comment by malcomm** [comment]
@ammarkarachi - I just commented out the the line and did a push. I am no longer blocked by this so not sure what happened exactly.

**Comment by ammarkarachi** [comment]
@malcomm Alright let me know if there is anything else, Thanks

**Comment by malcomm** [comment]
@ammarkarachi 

OK I just ran into a very similar issue. Here's the original schema:

```
type Foo
  @model
  @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
{
  id: ID!,
  studyEncounterSubjectId: ID!,
  ...
}
```

Adding a single key like this:

```
type Foo
  @model
  @key(name: "FooByUpdatedAt", fields: ["searchIndex", "updatedAt"], queryField: "fooByUpdatedAt")
  @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
{
  id: ID!,
  studyEncounterSubjectId: ID!,
  ...
}
```

Is resulting in:

```
Attempting to mutate more than 1 global secondary index at the same time on the FooTable table in the Foo stack. 
An error occurred during the push operation: Attempting to mutate more than 1 global secondary index at the same time on the FooTable table in the Foo stack. 
```

**Comment by dsypniewski** [comment]
I have the same problem, and I just found out that the order of the @key directives seems to matter.
Going from this:
```
type Foo
  @model
  @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
```
to this:
```
type Foo
  @model
  @key(name: "FooByUpdatedAt", fields: ["searchIndex", "updatedAt"], queryField: "fooByUpdatedAt")
  @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
```
results in abovementioned error, but instead doing:
```
type Foo
  @model
  @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
  @key(name: "FooByUpdatedAt", fields: ["searchIndex", "updatedAt"], queryField: "fooByUpdatedAt")
```
seems to work.
My guess is that the diffing in the first case creates something like this:
```
DELETE FooSubjectId
CREATE FooByUpdatedAt
CREATE FooSubjectId
```
instead of this:
```
CREATE FooByUpdatedAt
```

**Comment by malcomm** [comment]
@dsypniewski - thank you for the potential workaround. I was hopeful that it would work, but I tried messing around the order and I still got the error.

**Comment by dsypniewski** [comment]
@malcomm Have you tried to compare your schema.graphql with the one in #current-cloud-backend ? Those two files should be exactly the same except one line with the new `@key` after all the existing ones for that model.

**Comment by malcomm** [comment]
@dsypniewski - I just compared them and the only diff is the `@key`.

**Comment by malcomm** [comment]
@ammarkarachi - any updates on this? This is blocking a production fix.

**Comment by tsesamson** [comment]
@malcomm same issue for me and reverse the order of the @key didn't work for me.  Amplify will push it as long as I have a single @key.

**Comment by tsesamson** [comment]
Just figured out my issue was caused by the following field:
```
@key(name: "MessagesByAuthor", fields: ["authorId", "createdAt"], queryField: "messagesByAuthor")
...
author: User @connection(name: "MessagesByAuthor")
...
```
This caused amplify to build another index gsi-MessagesByAuthor on top of the @key directive I was trying to use.  This indeed was forcing amplify to create two gsi.  What I should have done was just define the field instead name for the author connection.

**Comment by mikerudge** [comment]
Getting the same problem. 
I am running amplify cli 4.29.8

I am trying to add-in
 
```
@key(
  name: "ActionLogsByCaseId"
  fields: ["actionLogCaseId"]
  queryField: "actionLogsByCaseId"
)
```

to this existing model... never had problems doing this before. I have also tried putting it in different positions within the schema but no luck. 

```
type ActionLog implements Meta
@model
@auth(
  rules: [
    { allow: owner }
    {
      allow: groups
      groupsField: "actionLogAccountId"
      operations: [create, update, read]
    }
    { allow: groups, groups: ["Admin"] }
  ]
)
@key(
  name: "ActionLogsByCompletedAt"
  fields: ["completedAt"]
  queryField: "actionLogsByCompletedAt"
)
@key(
  name: "ActionLogsByOwner"
  fields: ["owner", "createdAt"]
  queryField: "actionLogsByOwner"
)
@key(
  name: "ActionLogsByActionId"
  fields: ["actionLogActionId", "createdAt"]
  queryField: "actionLogsByActionId"
)
@key(
  name: "ActionLogsByAccountId"
  fields: ["actionLogAccountId", "createdAt"]
  queryField: "actionLogsByAccountId"
) {
  id: ID!
  completedAt: AWSDateTime
  case: Case! @connection(name: "ActionCaseLogs")
  actionLogCaseId: ID!
  owner: String
  user: LimitedUser @function(name: "getUser-${env}")
  action: Action! @connection(name: "ActionLogs")
  actionLogActionId: ID!
  account: Account! @connection(name: "AccountActionLogs")
  actionLogAccountId: ID!
  type: ActionLogType
  # Meta Data
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  deletedAt: AWSDateTime
}
```

**Comment by andrewbtp** [comment]
Also have this issue and I'm just trying to delete a single key.
```
@key(
  name: "CandidateByShortLink"
  fields: ["shortLink"]
) 
@key(
  name: "CandidateByOfficialLocation"
  fields: ["officialLocationId", "positionTitle"],
  queryField: "candidatesByOfficialLocation"
)
@key(
  name: "CandidateByVerificationStatus"
  fields: ["verification", "verificationDateTime"]
  queryField: "candidatesByVerificationStatus"
)
```

to

```
@key(
  name: "CandidateByOfficialLocation"
  fields: ["officialLocationId", "positionTitle"],
  queryField: "candidatesByOfficialLocation"
)
@key(
  name: "CandidateByVerificationStatus"
  fields: ["verification", "verificationDateTime"]
  queryField: "candidatesByVerificationStatus"
)
```

**Comment by tsesamson** [comment]
@andrewbtp do you have any table field connection to the "CandidateByShortLink" key?  
For example:
`fieldname: String action: @connection(name: "CandidateByShortLink")`

**Comment by andrewbtp** [comment]
@tsesamson Nope. I was able to remove the key by reverting to `@aws-amplify/cli@4.27.3`

**Comment by TQx8** [comment]
I had the same issue. After downgrading it worked @andrewbtp .

**Comment by oookoook** [comment]
Hello, I have the same issue using the `@aws-amplify/cli@4.31.1`.

I don't have much to offer in terms of clarifying this issue, I just wanted to let you know that it happens to other people as well.

I have an unused `@key` in the schema named `byDomainUrl`:

```
# TODO remove byDomainUrl key 
type Product 
@model
@key(name: "byDomain", fields: ["domainID"], queryField: "productsByDomain")
@key(name: "byDomainUrl", fields: ["domainID", "url"], queryField: "productsByDomainUrl")
@key(name: "byUrl", fields: ["url"], queryField: "productByUrl")
@key(name: "byDomainOriginalId", fields: ["domainID", "originalId"], queryField: "productByDomainOriginalId")
@key(name: "byDomainPopularity", fields: ["domainID", "popularity"], queryField: "productsByDomainPopularity")
@auth(rules: [{ allow: private, provider: iam }, { allow: private, operations:[read], provider: userPools }])
{
  id: ID!
  domainID: ID!
  domain: Domain @connection(fields: ["domainID"])
  url: String!
  name: String!
  description: String
  price: Float
  image: String
  originalImage: String!
  originalId: String!
  category: [String!]!
  popularity: Int
}
```

I can't remove the `byDomainUrl` key, getting the same error message: `Attempting to mutate more than 1 global secondary index at the same time on the ProductTable table in the Product stack. An error occurred during the push operation: Attempting to mutate more than 1 global secondary index at the same time on the ProductTable table in the Product stack.`

The schema (and the DynamoDB GSI index) was created with one of the previous Amplify CLI versions, but I don't know which. I update regularly, and I started working on this project at the beginning of August.

I don't use the `@key` in any `@connection`.

**Comment by maximelebastard** [comment]
I'm having the same issue using a custom CloudFormation resource containing a DynamoDB table.
I'm only removing a previously created index from the file and Amplify doesn't want to push.

Downgrading to `@aws-amplify/cli@4.27.3` fixes it

**Comment by jaimeburnap-fts** [comment]
I am also having this error when removing a key. And the only way to add a key is to put it at the end of the existing keys.

**Comment by malcomm** [comment]
Any updates on this?

**Comment by Nxtra** [comment]
> I have the same problem, and I just found out that the order of the `@key` directives seems to matter.
> Going from this:
> 
> ```
> type Foo
>   @model
>   @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
> ```
> 
> to this:
> 
> ```
> type Foo
>   @model
>   @key(name: "FooByUpdatedAt", fields: ["searchIndex", "updatedAt"], queryField: "fooByUpdatedAt")
>   @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
> ```
> 
> results in abovementioned error, but instead doing:
> 
> ```
> type Foo
>   @model
>   @key(name: "FooSubjectId", fields: ["fooSubjectId"], queryField: "fooSubjectId")
>   @key(name: "FooByUpdatedAt", fields: ["searchIndex", "updatedAt"], queryField: "fooByUpdatedAt")
> ```
> 
> seems to work.
> My guess is that the diffing in the first case creates something like this:
> 
> ```
> DELETE FooSubjectId
> CREATE FooByUpdatedAt
> CREATE FooSubjectId
> ```
> 
> instead of this:
> 
> ```
> CREATE FooByUpdatedAt
> ```

I can confirm this workaround worked for me.
The order of the keys in my schema mattered.

**Comment by simon-lanf** [comment]
I have this problem right now. on 4.32.1.

I'm trying to add the first key on a model. It keeps saying the same error.

**Comment by Nxtra** [comment]
> I have this problem right now. on 4.32.1.
> 
> I'm trying to add the first key on a model. It keeps saying the same error.

Yes, I ran into it with the same version of the CLI too.
For me it was fixed by making the `@key` that I was changing / updating the last key added on top of the type in the graphql specification.

**Comment by oookoook** [comment]
Seems to be fixed (at least for me) in 4.36.2.

**Comment by frankclaassen** [comment]
This is still an issue in 4.36.2. Trying to delete a @key in the middle of @key list throws the error

**Comment by frankclaassen** [comment]
With @aws-amplify/cli@4.27.3 you are able to remove @key definitions, so clearly someone changed something and didn't do their testing properly....

**Comment by gildniy** [comment]
I used the @maximelebastard downgrade solution and it worked, and once pushed, I had to reinstall the latest version of the cli and continue

**Comment by malcomm** [comment]
Any updates? This is blocking production fix that is really needed.

@mikeparisstuff - any insight into this?

**Comment by jhockett** [comment]
You can enable a preview of iterative updates by updating the CLI and changing the value of the feature flag in `amplify/cli.json`. This feature is new and may not be suitable for a production app.

https://docs.amplify.aws/cli/reference/feature-flags#enableIterativeGsiUpdates

**Comment by mikeRChambers610** [comment]
I cant event update any of my @models with any new GSI or delete any old GSI because of this issue.

I am not sure what to do at this point.

**Comment by jcbdev** [comment]
@mikeRChambers610 @malcomm  I have this issue frequently.  Manually delete them in the aws console first.  Tends to work for me 🤷

**Comment by jacobdubail** [comment]
> @mikeRChambers610 @malcomm I have this issue frequently. Manually delete them in the aws console first. Tends to work for me 🤷

Are you suggesting to delete them in DynamoDB or AppSync? I assume in DynamoDB since attempting to do so in AppSync throws a new flavor of error.

**Comment by warrenmcquinn** [comment]
I'm still using Amplify 4.27.3 because this issue has not been resolved.

**Comment by Reklino** [comment]
@warrenmcquinn I had this issue on 4.32.3 and updated to 4.44.2 and it fixed it for me. Might be worth trying the latest version.

**Comment by vincent38wargnier** [comment]
I have the same issue, it seems that we can't update or create 2 @key of the same table at the same push.

So you will need to push them one by one. What is pretty funny (my case) is that this push I have to do is due to merging master with my dev env. It means that between the several push I have to do due to this bug, my production web site will be fucked up. 
Thanks again amazing aws

**Comment by kldeb** [comment]
You have to work with the limitations of the tools you use and determine if all the other stuff it does for you is worth it. 

We release our changes with one key change at a time. Or we work on separate branches and merge changes in sequence.

**Comment by vincent38wargnier** [comment]
Yes that's what I did. But if tomorrow I create a new branch, that is a duplicate of my actual branch, i will need to go through this procedure for every table that has 2 @key or more. That's not normal. What will I say to my employee that needs to duplicate the master? And what if i have 3000 tables with multiple @keys? 
Your solution has no cense. It works only for small school project, not for pro ones

**Comment by josefaidt** [comment]
Hey y'all this has since been released and is enabled by default 🙂 https://aws.amazon.com/about-aws/whats-new/2021/04/aws-amplify-orchestrates-multiple-amazon-dynamodb-gsi-updates-single-deployment/
